### PR TITLE
Added the possibility to cancel research.

### PR DESCRIPTION
**Pull request by MortenTobiasNielsen**

### CODE REVIEW

**Comment by theis999** [comment]
I am confused, what is this supposed to do?

**Comment by MortenTobiasNielsen** [comment]
It cancels the current research and starts a new if the comment for a Tech step is _Cancel_.

So if I have started Steelaxe due to a lack of Logistics Science Packs and then want to cancel it and start Advanced Material Processing, then I can make an _Advanced Material Processing_ tech step with _Cancel_ as the comment. A completely made up scenario and totally not the reason I had to make the change. :D

File: Lua Files/control.lua [hunk 1]
```diff
@@ -853,8 +853,17 @@ local function recipe()
 end
 
 local function tech()
+	if steps[step].comment and steps[step].comment == "Cancel" then
+		if player.force.current_research then
+			msg(string.format("Cancel research %s and add research %s", player.force.current_research.name, item))
+			player.force.cancel_current_research();
```
Comment (theis999) [inline Lua Files/control.lua:859]: We usually don't use semicolons in lua 😉

**Comment by theis999** [comment]
Alternatively to wiping the whole research queue. The cancel part should return false, so it cost a tick per item that is cancelled before paying the tick to queue the research you want.

File: Lua Files/control.lua [hunk 2]
```diff
@@ -853,8 +853,17 @@ local function recipe()
 end
 
 local function tech()
+	if steps[step].comment and steps[step].comment == "Cancel" then
+		if player.force.current_research then
+			msg(string.format("Cancel research %s and add research %s", player.force.current_research.name, item))
```
[the remark this answers is not in the document]
Comment (MortenTobiasNielsen) [reply Lua Files/control.lua:858]: Changed to "Research: %s cancelled and %s added"

File: Lua Files/control.lua [hunk 3]
```diff
@@ -853,8 +853,17 @@ local function recipe()
...
+			player.force.cancel_current_research();
+			player.force.add_research(item)
+			return true
+		end
+	end
+
 	player.force.add_research(item)
```
[the remark this answers is not in the document]
Comment (MortenTobiasNielsen) [reply Lua Files/control.lua:865]: I usually like not putting code into else blocks if it isn't needed. :)

File: Lua Files/control.lua [hunk 4]
```diff
@@ -853,8 +853,17 @@ local function recipe()
...
-    msg(string.format("Add research %s", item))
+	msg(string.format("Add research %s", item))
```
[the remark this answers is not in the document]
Comment (MortenTobiasNielsen) [reply Lua Files/control.lua:866]: Agreed. I have changed it so it is the same for both.

**Comment by MortenTobiasNielsen** [comment]
I have changed it to player.force.research_queue, that does seem to make more sense and I have changed most of what you have pointed out. :)

**Review by theis999 — APPROVED** [review]

**Pull request merged**